File: scripts/news.py
"""Headlines for Danny's own players, from ESPN's public NFL news feed (no key).

The Node side is lib/news.js; this is the same join, for the offline bake. Sleeper only carries a
"news updated" date, never the story, so the text comes from ESPN and is matched to the roster on
Sleeper's espn_id, with a name match as the fallback. Best effort: a failure leaves the list empty.
"""
import json
import urllib.request
from concurrent.futures import ThreadPoolExecutor

FEED = "https://site.api.espn.com/apis/site/v2/sports/football/nfl/news?limit=50"
TEAM_FEED = "https://site.api.espn.com/apis/site/v2/sports/football/nfl/news?limit=20&team={}"


def _get(url, timeout=8):
    req = urllib.request.Request(url, headers={"User-Agent": "robinsavages"})
    with urllib.request.urlopen(req, timeout=timeout) as r:
        return json.load(r)


def player_news(rows):
    """rows: [{id, name, team, espn_id}] -> (stories newest first, error or None)."""
    if not rows:
        return [], None
    teams = sorted({r.get("team") for r in rows if r.get("team") and r["team"] != "FA"})
    urls = [FEED] + [TEAM_FEED.format(t) for t in teams]

    def pull(u):
        try:
            return _get(u)
        except Exception as exc:  # noqa: BLE001
            return {"_error": str(exc)}

    with ThreadPoolExecutor(max_workers=6) as pool:
        feeds = list(pool.map(pull, urls))
    failed = [f for f in feeds if f.get("_error")]
    if len(failed) == len(feeds):
        return [], failed[0]["_error"]

    articles, seen = [], set()
    for f in feeds:
        for a in f.get("articles") or []:
            key = ((a.get("links") or {}).get("web") or {}).get("href") or a.get("headline")
            if key and key not in seen:
                seen.add(key)
                articles.append(a)

    by_espn = {str(r["espn_id"]): r for r in rows if r.get("espn_id")}
    named = [r for r in rows if " " in (r.get("name") or "")]

    stories = []
    for a in articles:
        hit = {}
        for c in a.get("categories") or []:
            ath = c.get("athlete") or {}
            if ath.get("id") is not None and str(ath["id"]) in by_espn:
                r = by_espn[str(ath["id"])]
                hit[r["id"]] = r
        text = f"{a.get('headline') or ''} {a.get('description') or ''}"
        for r in named:
            if r["id"] not in hit and r["name"] in text:
                hit[r["id"]] = r
        if not hit:
            continue
        stories.append({"headline": a.get("headline") or "", "summary": a.get("description") or "",
                        "published": a.get("published"), "published_ms": _ms(a.get("published")),
                        "link": ((a.get("links") or {}).get("web") or {}).get("href"),
                        "players": [{"id": r["id"], "name": r["name"]} for r in hit.values()]})
    stories.sort(key=lambda s: s["published_ms"] or 0, reverse=True)
    return stories, (f"{len(failed)} of {len(feeds)} ESPN feeds failed" if failed else None)


def _ms(iso):
    if not iso:
        return None
    try:
        import datetime
        return int(datetime.datetime.fromisoformat(iso.replace("Z", "+00:00")).timestamp() * 1000)
    except Exception:  # noqa: BLE001
        return None

File: lib/news.js
// Headlines for Danny's own players.
// Sleeper only carries a "news updated" timestamp, never the story, so the text comes from ESPN's public
// NFL news feed (no key, same best-effort contract as lib/games.js: a failure leaves the list empty).
// The join is Sleeper's espn_id against the feed's athlete categories, with a name match as the fallback
// for stories ESPN tagged loosely.

const FEED = "https://site.api.espn.com/apis/site/v2/sports/football/nfl/news?limit=50";
// the league-wide feed alone rarely mentions a given roster; ESPN also filters the same feed by team
const TEAM_FEED = t => `https://site.api.espn.com/apis/site/v2/sports/football/nfl/news?limit=20&team=${t}`;

async function fetchJSON(url, ms = 8000) {
  const ctl = new AbortController();
  const t = setTimeout(() => ctl.abort(), ms);
  try {
    const r = await fetch(url, { signal: ctl.signal, headers: { "User-Agent": "robinsavages" } });
    if (!r.ok) throw new Error(`${r.status} ${r.statusText}`);
    return await r.json();
  } finally { clearTimeout(t); }
}

/**
 * @param {Array<{id:string,name:string,espn_id?:string|number,team?:string}>} rows the players to match
 * @returns {Promise<{stories:Array,error:string|null}>} newest first; every story names the players it is about
 */
export async function playerNews(rows = []) {
  if (!rows.length) return { stories: [], error: null };
  const teams = [...new Set(rows.map(r => r.team).filter(t => t && t !== "FA"))];
  const feeds = await Promise.all([FEED, ...teams.map(TEAM_FEED)].map(u => fetchJSON(u).catch(e => ({ error: String(e.message || e) }))));
  const failed = feeds.filter(f => f.error);
  if (failed.length === feeds.length) return { stories: [], error: failed[0].error };
  const articles = [], seen = new Set();
  for (const f of feeds) for (const a of (f.articles || [])) {
    const key = ((a.links || {}).web || {}).href || a.headline;
    if (key && !seen.has(key)) { seen.add(key); articles.push(a); }
  }
  const feed = { articles };

  const byEspn = {};
  for (const r of rows) if (r.espn_id) byEspn[String(r.espn_id)] = r;
  const named = rows.filter(r => (r.name || "").includes(" "));

  const stories = [];
  for (const a of (feed.articles || [])) {
    const cats = a.categories || [];
    const hit = new Map();
    for (const c of cats) {
      const id = c.athlete && c.athlete.id != null ? String(c.athlete.id) : null;
      if (id && byEspn[id]) hit.set(byEspn[id].id, byEspn[id]);
    }
    const text = `${a.headline || ""} ${a.description || ""}`;
    for (const r of named) if (!hit.has(r.id) && text.includes(r.name)) hit.set(r.id, r);
    if (!hit.size) continue;
    const ms = Date.parse(a.published || a.lastModified || "") || null;
    stories.push({
      headline: a.headline || "", summary: a.description || "",
      published: a.published || null, published_ms: ms,
      link: ((a.links || {}).web || {}).href || null,
      players: [...hit.values()].map(r => ({ id: r.id, name: r.name })),
    });
  }
  stories.sort((a, b) => (b.published_ms || 0) - (a.published_ms || 0));
  return { stories, error: failed.length ? `${failed.length} of ${feeds.length} ESPN feeds failed` : null };
}
